carmm/build/mixed_oxide.py

```diff
@@ -0,0 +1,98 @@
+    :param metal1_moment: The initial moment on metal A that you have specified in the input geometry.in
+    :param metal2_moment:
+    :param up_or_down_spin: if the metal1 has zero magnetic moment and metal2 has a magnetic moment specified by
+    metal2_moment, please specify the sign for the magnetic moment of the metal2. Positive (up-spin) or Negative (down-spin)
+    :return:
+    '''
+    if up_or_down_spin!=None:
+        up_or_down_spin=up_or_down_spin.capitalize()
+        print(up_or_down_spin)
+
+    atoms = atoms_object
+    sym = np.array(atoms.get_chemical_symbols())
+    charge= np.array(atoms.get_initial_charges())
+    mom = np.array(atoms.get_initial_magnetic_moments())
+    mom_sign= np.sign(mom)
+    m1_ind = []
+    for ind,symbol in enumerate(sym):
+        if symbol==metal1.capitalize() and charge[ind]==metal1_charge:
+            if np.abs(mom[ind])==metal1_moment:
+                m1_ind.append(ind)
+
+    # print(m1_ind)
+    # determining the number of metal1 ions to be replaced with metal2 which depends on the ratio provided by the user.
+    k = math.floor((len(m1_ind)/(1+ratio)))
+    # random replacement of metal1 ion with metal2 ions
+    m2_ind = random.sample(m1_ind, k)
+    '''
+    This functionality relies on the random replacement of metal 1 ions with metal 2 ions. Strictly speaking, randomly 
+    generated geometry files for a given ratio must be tested for energy and it should be figured out which one is the 
+    most stable. 
+    But this combinatorial problem can be very huge and hence expensive.'''
+    # print(m2_ind)
+    for i in m2_ind:
+        sym[i] = metal2.capitalize()
+        charge[i] = metal2_charge
+        if mom_sign[ind]==0:
+            if up_or_down_spin==None:
+                raise Exception('the magnetic moment of the metal atom that you desire to replace has is zero'
+                                '(non-magnetic). Please provide the value for the argument up_or_down_spin')
+            elif up_or_down_spin=='Positive':
+                print('the magnetic moment of the metal atom that you desire to replace is zero(non-magnetic). '
+                      'Using a positive (up-spin) moment')
+                mom[i] = metal2_moment
+            elif up_or_down_spin == 'Negative':
+                print('the magnetic moment of the metal atom that you desire to replace has is zero(non-magnetic). '
+                      'Using a negative (down-spin) moment')
+                mom[i] = (-1)*metal2_moment
+            else:
+                raise Exception(f'Invalid input to the argument up_or_down_spin. Should be either \'positive\' or '
+                                f'\'negative\'. You provided up_or_down_spin={up_or_down_spin}')
+        elif mom_sign[ind]!=0:
+            mom[i] = mom_sign[ind]*metal2_moment
+
+    atoms.set_chemical_symbols(sym)
+    atoms.set_initial_charges(charge)
+    atoms.set_initial_magnetic_moments(mom)
+
+    if neutrality_check(atoms):
```

Comment:
Why have you got a function here that is writing a file? I'm confused - we shouldn't have any read/write functionality in the functions, only passing atoms objects in/out. We don't want to shield users from managing an atoms object (unless it is a workflow)